core/src/main/java/jenkins/security/ResourceDomainRootAction.java

```diff
@@ -0,0 +1,270 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.security;
+
+import hudson.Extension;
+import hudson.Util;
+import hudson.model.DirectoryBrowserSupport;
+import hudson.model.UnprotectedRootAction;
+import hudson.model.User;
+import hudson.security.ACL;
+import hudson.security.ACLContext;
+import hudson.security.AccessControlled;
+import jenkins.model.Jenkins;
+import jenkins.util.SystemProperties;
+import org.acegisecurity.AccessDeniedException;
+import org.acegisecurity.Authentication;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.*;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Date;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Root action serving {@link DirectoryBrowserSupport} instances
+ * on random URLs to support resource URLs (second domain).
+ *
+ * @see ResourceDomainConfiguration
+ * @see ResourceDomainFilter
+ *
+ * @since TODO
+ */
+@Extension
+@Restricted(NoExternalUse.class)
+public class ResourceDomainRootAction implements UnprotectedRootAction {
+
+    private static final Logger LOGGER = Logger.getLogger(ResourceDomainRootAction.class.getName());
+
+    @CheckForNull
+    @Override
+    public String getIconFileName() {
+        return null;
+    }
+
+    @CheckForNull
+    @Override
+    public String getDisplayName() {
+        return null;
+    }
+
+    @CheckForNull
+    @Override
+    public String getUrlName() {
+        return "static-files";
+    }
+
+    public void doIndex(StaplerRequest req, StaplerResponse rsp) throws IOException {
+        if (!ResourceDomainConfiguration.isResourceRequest(req)) {
+            rsp.sendError(404, "Cannot handle requests to this URL unless on Jenkins resource URL.");
+        } else {
+            rsp.sendError(404, "Jenkins serves only static files on this domain.");
+        }
+    }
+
+    public Object getDynamic(String id, StaplerRequest req, StaplerResponse rsp) throws Exception {
+        if (!ResourceDomainConfiguration.isResourceRequest(req)) {
+            rsp.sendError(404, "Cannot handle requests to this URL unless on Jenkins resource URL.");
+            return null;
+        }
+
+        Token token = Token.decode(id);
+        if (token == null) {
+            rsp.sendError(404, "Jenkins serves only static files on this domain.");
+            return null;
+        }
+
+        String authenticationName = token.username;
+        String browserUrl = token.path;
+
+        long creationDate = token.timestamp;
+        long age = new Date().getTime() - creationDate;
+
+        if (age >= 0 && age < TimeUnit.MINUTES.toMillis(VALID_FOR_MINUTES)) {
```

Comment:
Nit: you can use `java.time.Duration` and `Instant` here for a more fluent API.

Reply:
The best I could come up with (including static imports) is

```
Instant creationDate = ofEpochMilli(token.timestamp);
if (creationDate.plus(VALID_FOR_MINUTES, MINUTES).isAfter(now()) && creationDate.isBefore(now())) {
```

which seems like it's not much of an improvement…

Reply:
Yeah, guess there's nothing to do here.

Reply:
Did it anyway, minimally nicer but since I had it already…